src/gc/gc.cpp

```diff
@@ -34233,6 +34236,20 @@ HRESULT GCHeap::Initialize()
             {
                 pmask &= smask;
 
+#ifdef FEATURE_PAL
+                // GetCurrentProcessAffinityMask can return pmask=0 and smask=0 on
```

Comment:
do we know why this is?

Reply:
I've implemented that based on the MSDN doc, which says:
> On a system with more than 64 processors, if the threads of the calling process are in a single processor group, the function sets the variables pointed to by lpProcessAffinityMask and lpSystemAffinityMask to the process affinity mask and the processor mask of active logical processors for that group. If the calling process contains threads in multiple groups, **the function returns zero for both affinity masks.**

Reply:
So I think it can happen on Windows too.

Reply:
this is on the path where `if (!(GCToOSInterface::CanEnableGCCPUGroups()))` so we are saying there's < 64 procs.

Reply:
On my 96-core machine, this API returns a pmask set to 48 cores on Windows & 0 on Linux. For an 8-core machine, we see a pmask set to 8 cores on both Windows/Linux.

Reply:
> so we are saying there's < 64 procs.

That's not necessarily true. The GCToOSInterface::CanEnableGCCPUGroups can return FALSE even if there are more than 64 processors when COMPlus_GCCpuGroup=0.

Reply:
I meant there are < 64 procs **available to this process.** the doc says if there are more than 64 procs but if the calling process uses < 64 procs, this function is still returning valid pmask/smask on windows.

Reply:
I understand, but I was trying to say that even if there are > 64 processors available to this process, we can still get to this code path if the COMPlus_GCCpuGroup=0.

Reply:
Actually, I am starting to doubt I understand what the _If the calling process contains threads in multiple groups_ in the MSDN doc means. I have read it as "if the current process has affinity mask set to multiple groups" and that's how I have implemented it in the PAL.

Reply:
you are right - if you have GCCpuGroup set to 0 it would be more than 64 procs available to this process while CanEnableGCCCpuGroups is FALSE.

I'm looking at the cpu group code in util.cpp. the policy it uses is a little odd to me - it enables cpu groups by default instead of checking to see if one of the configs wants to enable cpu groups and then enable it if that's the case. but by default processes do not use more than one cpu group worth of processors. what's the policy on linux? if you have > 64 procs do processes use all procs by default? 

there is a discrepancy on windows and linux regardless, we should unify the behavior.

Reply:
> if you have > 64 procs do processes use all procs by default?

Linux has a very different way of reporting / setting affinity and there is no special handling for more or less than 64 processors. It doesn't have any groups. These are all Windows specific constructs.

There is `sched_getaffinity` / `sched_setaffinity` that use a `cpu_set_t` which can be manipulated as described here: https://linux.die.net/man/3/cpu_set. It is implemented as a bitset that can hold as many bits as needed for all the processors in the system.
Then there is a function `numa_node_to_cpus` that fills in a `cpu_set_t` with all processors belonging to the requested numa node index where the numa node index can be a value from 0 to `numa_max_node() - 1` And finally `numa_num_possible_cpus()` that returns the number of cpus enabled by the kernel (there is a kernel option that allows you to limit that number at boot time if needed).

Reply:
So my code in PAL takes these values and transforms them into the Windows style, artificially creating groups so that processors in a group belong to single NUMA node. So e.g. on my box, I have two NUMA nodes each containing 4 CPUs. So I create two groups with 4 processors each.